File: Problems/Hackerrank/EpicCode/6.cpp
/* 
* @Author: andreshp
* @Date:   2015-06-21
* @Last Modified by:   andreshp
* @Last Modified time: 2015-06-21
*/

#include <iostream>

using namespace std;

const int MAX_N = 200000;
long array[MAX_N];

//---------------------------- FUNCTIONS ----------------------------//

void query1(long array[], int x, int y, int N){
    long count = 0;
    for (int i = 0; i < y-x+1; i++){
        count = (count + (i+1)*(i+2)) % 1000000007;
        array[x+i] = (array[x+i] + count) % 1000000007;        
    }

    for (int i = y+1; i < N; i++)
        array[i] = (array[i] + count) % 1000000007;    
}

long query2(int array[], int x, int y){
    if (x == 0)
        return array[y];
    else{
        long sol = array[y] - array[x-1];
        if (sol < 0)
            sol = 1000000007 + sol;
        return sol;
    }
}

//------------------------------ MAIN -------------------------------//

int main(){
    int N, Q, t, x, y;
    cin >> N >> Q;
    for (int i = 0; i < N; i++)
        array[i] = 0;

    for (int q = 0; q < Q; q++){
        cin >> t >> x >> y;
        if (t == 1)
            query1(array,x-1,y-1, N);
        else
            cout << query2(array,x-1,y-1);
    }

    return 0;
}

File: Problems/Hackerrank/EpicCode/6.py
#!/usr/bin/env python

###################################################################
# Author: andreshp
# Email:  andreshp9@gmail.com
# Date:   2015-06-21
# Last Modified by:   andreshp
# Last Modified time: 2015-06-21
# File Name: 6.py
###################################################################

import math
import sys
import time

#---------------------------- FUNCTIONS ----------------------------#

def query1(array, x, y):
    count = 0
    for i in range(0, y-x+1):
        count = (count + (i+1)*(i+2)) % 1000000007
        array[x+i] = (array[x+i] + count) % 1000000007

    for i in range(y+1, len(array)):
        array[i] = (array[i] + count) % 1000000007

def query2(array,x,y):
    if x == 0:
        return array[y]
    else:
        sol = array[y] - array[x-1]
        if sol < 0:
            sol = 1000000007 + sol
        return sol

#------------------------------ MAIN -------------------------------#

N, Q = [int(x) for x in input().split()]
array = [0 for i in range(0,N)]
for i in range(0,Q):
    t, x, y = [int(x) for x in input().split()]
    if t == 1:
        query1(array,x-1,y-1)
    else:
        print(query2(array,x-1,y-1))